token-lending/program/src/processor.rs

```diff
@@ -1874,6 +1940,57 @@ fn process_flash_loan(
     Ok(())
 }
 
+#[inline(never)] // avoid stack frame limit
+fn process_withdraw_obligation_collateral_and_redeem_reserve_liquidity(
+    program_id: &Pubkey,
+    collateral_amount: u64,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let reserve_collateral_info = next_account_info(account_info_iter)?;
+    let user_collateral_info = next_account_info(account_info_iter)?;
+    let reserve_info = next_account_info(account_info_iter)?;
+    let obligation_info = next_account_info(account_info_iter)?;
+    let lending_market_info = next_account_info(account_info_iter)?;
+    let lending_market_authority_info = next_account_info(account_info_iter)?;
+    let user_liquidity_info = next_account_info(account_info_iter)?; 
+    let reserve_collateral_mint_info = next_account_info(account_info_iter)?; 
+    let reserve_liquidity_supply_info = next_account_info(account_info_iter)?; 
+    let obligation_owner_info = next_account_info(account_info_iter)?;
+    let user_transfer_authority_info = next_account_info(account_info_iter)?; 
+    let clock = &Clock::from_account_info(next_account_info(account_info_iter)?)?;
+    let token_program_id = next_account_info(account_info_iter)?;
+
+    let liquidity_amount = _withdraw_obligation_collateral(
+        program_id, 
+        collateral_amount, 
+        reserve_collateral_info, 
+        user_collateral_info, 
+        reserve_info, 
+        obligation_info, 
+        lending_market_info, 
+        lending_market_authority_info, 
+        obligation_owner_info, 
+        clock, 
+        token_program_id,
+    )?;
+
```

Comment:
So we don't have to refresh anything here because _withdraw_obligation_collateral requires the reserve to be refreshed, and does not mark it as stale (only reads from it as far as I can tell)

Reply:
oh hm maybe my flow chart is wrong then

Reply:
Its not strictly wrong, since it *does* require a refreshed reserve.
If someone were to do these in separate transactions, you might need to re-refresh it.

Reply:
i guess it makes sense since removing the collateral doesn't change anything about the deposited liquidity so no interest calculations need to happen

Reply:
Just got around to checking this, it does seem like it needs to be refreshed?

![image](https://user-images.githubusercontent.com/83670135/125406450-6ec29e00-e3eb-11eb-99a3-aa1829050931.png)
`_withdraw_obligation_collateral` marks stale

![image](https://user-images.githubusercontent.com/83670135/125406491-7b46f680-e3eb-11eb-8bbc-90bedc08854d.png)
`_redeem_reserve_collateral` checks if stale

Reply:
Thats just the obligation thats getting marked stale right?

Reply:
ye

Reply:
Did the frontend PR, works like a charm! https://github.com/solendprotocol/web/pull/114